src/VirtoCommerce.CatalogModule.Data/Validation/CategoryPropertyNameValidator.cs

```diff
@@ -1,36 +1,122 @@
 
 namespace VirtoCommerce.CatalogModule.Data.Validation
 {
     public class CategoryPropertyNameValidator : AbstractValidator<CategoryPropertyValidationRequest>
     {
-        private readonly IPropertySearchService _propertySearchService;
+        private readonly ICategoryService _categoryService;
+        private readonly IPropertyService _propertyService;
+        private readonly ICategoryIndexedSearchService _categoryIndexedSearchService;
+        private readonly Func<ICatalogRepository> _catalogRepositoryFactory;
+        private readonly ISettingsManager _settingsManager;
 
-        public CategoryPropertyNameValidator(IPropertySearchService propertySearchService)
+        public CategoryPropertyNameValidator(
+            ICategoryService categoryService,
+            IPropertyService propertyService,
+            ICategoryIndexedSearchService categoryIndexedSearchService,
+            Func<ICatalogRepository> catalogRepositoryFactory,
+            ISettingsManager settingsManager
+            )
         {
-            _propertySearchService = propertySearchService;
+            _categoryService = categoryService;
+            _propertyService = propertyService;
+            _categoryIndexedSearchService = categoryIndexedSearchService;
+            _catalogRepositoryFactory = catalogRepositoryFactory;
+            _settingsManager = settingsManager;
+
             AttachValidators();
         }
 
         private void AttachValidators()
         {
-            RuleFor(r => r).MustAsync(async (r, _) =>
-            {
-                var propertySearchCriteria = new PropertySearchCriteria
+            RuleFor(r => r)
+                .CustomAsync(async (r, context, _) =>
                 {
-                    PropertyNames = new[] { r.PropertyName },
-                    CategoryId = r.CategoryId
-                };
-                var propertySearchResult = await _propertySearchService.SearchPropertiesAsync(propertySearchCriteria);
+                    (var isPropertyUniqueForHierarchy, var CategoryName) = await CheckPropertyUniquenessAsync(r);
+
+                    if (!isPropertyUniqueForHierarchy)
+                    {
+                        context.AddFailure(new ValidationFailure(r.PropertyName, "duplicate-property")
+                        {
+                            CustomState = new
+                            {
+                                PropertyName = r.PropertyName,
+                                CategoryName = CategoryName,
+                            },
+                        });
+                    }
+                });
+        }
+
+        protected virtual async Task<(bool, string)> CheckPropertyUniquenessAsync(CategoryPropertyValidationRequest request)
+        {
+            var useIndexedSearch = await _settingsManager.GetValueAsync(catalogCore.ModuleConstants.Settings.Search.UseCatalogIndexedSearchInManager.Name, true);
+
+            var categoryIds = new List<string>();
```

Comment:
Can we move this block to the protected method?
![image](https://user-images.githubusercontent.com/47894313/147245041-6640c69f-7e6e-4937-929e-cbbfa66719df.png)